File: Session_10_StringMatching/multi_language_implementations/rabin_karp.kt
const val d = 256
const val q = 101

fun runRabinKarp(pat: String, txt: String) {
    val M = pat.length
    val N = txt.length
    var p = 0
    var t = 0
    var h = 1

    for (i in 0 until M - 1) {
        h = (h * d) % q
    }

    for (i in 0 until M) {
        p = (d * p + pat[i].code) % q
        t = (d * t + txt[i].code) % q
    }

    println("Rabin-Karp Search details:")
    println("  Pattern hash: $p")

    for (i in 0..N - M) {
        if (p == t) {
            println("  Hash match found at index $i (hash: $t). Checking characters...")
            var j = 0
            while (j < M) {
                if (txt[i + j] != pat[j]) {
                    break
                }
                j++
            }
            if (j == M) {
                println("    Pattern found at index: $i")
            }
        }

        if (i < N - M) {
            t = (d * (t - txt[i].code * h) + txt[i + M].code) % q
            if (t < 0) {
                t += q
            }
        }
    }
}

fun main() {
    val txt = "GEEKS FOR GEEKS"
    val pat = "GEEK"

    println("=== Rabin-Karp String Search ===")
    println("Text: \"$txt\"")
    println("Pattern: \"$pat\"\n")

    runRabinKarp(pat, txt)

    println()
    print("Press Enter to exit...")
    readLine()
}

File: Session_10_StringMatching/multi_language_implementations/rabin_karp.js
const readline = require('readline');

const d = 256;
const q = 101;

function runRabinKarp(pat, txt) {
    const M = pat.length;
    const N = txt.length;
    let p = 0;
    let t = 0;
    let h = 1;

    for (let i = 0; i < M - 1; i++) {
        h = (h * d) % q;
    }

    for (let i = 0; i < M; i++) {
        p = (d * p + pat.charCodeAt(i)) % q;
        t = (d * t + txt.charCodeAt(i)) % q;
    }

    console.log("Rabin-Karp Search details:");
    console.log(`  Pattern hash: ${p}`);

    for (let i = 0; i <= N - M; i++) {
        if (p === t) {
            console.log(`  Hash match found at index ${i} (hash: ${t}). Checking characters...`);
            let j = 0;
            for (j = 0; j < M; j++) {
                if (txt[i + j] !== pat[j]) {
                    break;
                }
            }
            if (j === M) {
                console.log(`    Pattern found at index: ${i}`);
            }
        }

        if (i < N - M) {
            t = (d * (t - txt.charCodeAt(i) * h) + txt.charCodeAt(i + M)) % q;
            if (t < 0) {
                t += q;
            }
        }
    }
}

function main() {
    const txt = "GEEKS FOR GEEKS";
    const pat = "GEEK";

    console.log("=== Rabin-Karp String Search ===");
    console.log(`Text: "${txt}"`);
    console.log(`Pattern: "${pat}"\n`);

    runRabinKarp(pat, txt);

    console.log();
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();
